File: problems/HackerRank/2521_exercises/tree/isBalanced.c
struct Result
{
	bool isBalanced;
	int height;
};

int max(int a, int b)
{
	return a > b ? a : b;
}

struct Result bfs(struct TreeNode *root)
{
	if (root == NULL)
	{
		struct Result result = {true, 0};
		return result;
	}

	struct Result left = bfs(root->left);
	struct Result right = bfs(root->right);

	bool balance = left.isBalanced && right.isBalanced && (abs(left.height - right.height) <= 1);

	struct Result result = {balance, 1 + max(left.height, right.height)};

	return result;
}

bool isBalanced(struct TreeNode *root)
{
	return bfs(root).isBalanced;
}

File: problems/LeetCode/tree/isBalanced.py
def isBalanced(self, root: Optional[TreeNode]) -> bool:

		def getHeight(node: Optional[TreeNode] ) -> int:
				if not node:
						return 0

				left_height = getHeight(node.left)

				if left_height == -1:
						return -1
				
				right_height = getHeight(node.right)

				if right_height == -1 or abs(left_height - right_height) > 1:
						return -1
				
				return max(left_height, right_height) + 1
		
		return getHeight(root) != -1


def isBalanced(self, root: Optional[TreeNode]) -> bool:

		def dfs(node: Optional[TreeNode]):
				if not node:
						return [True, 0]

				left, right = dfs(node.left), dfs(node.right)

				balanced = left[0] and right[0] and abs(left[1] - right[1]) <= 1

				return [balanced, max(left[1],right[1]) + 1]

		return dfs(root)[0]